pkg/provisioner/ironic/ironic.go

```diff
@@ -794,6 +779,85 @@ func (p *ironicProvisioner) Deprovision(deleteIt bool) (result provisioner.Resul
 	}
 }
 
+// Delete removes the host from the provisioning system. It may be
+// called multiple times, and should return true for its dirty flag
+// until the deprovisioning operation is completed.
+func (p *ironicProvisioner) Delete() (result provisioner.Result, err error) {
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+	if ironicNode == nil {
+		p.log.Info("no node found, already deleted")
+		return result, nil
+	}
+
+	p.log.Info("deleting host",
+		"ID", ironicNode.UUID,
+		"lastError", ironicNode.LastError,
+		"current", ironicNode.ProvisionState,
+		"target", ironicNode.TargetProvisionState,
+		"deploy step", ironicNode.DeployStep,
+	)
+
+	if ironicNode.ProvisionState == nodes.Available {
+		// Move back to manageable so we can delete it cleanly.
+		return p.changeNodeProvisionState(
+			ironicNode,
+			nodes.ProvisionStateOpts{Target: nodes.TargetManage},
+		)
+	} else if !ironicNode.Maintenance {
```

Comment:
This will try to set maintenance for nodes in `manageable` state. I think you missed something like `ironicNode.ProvisionState != nodes.Manageable`. Well, you can also delete nodes in `enroll` state...

Reply:
Does setting the flag in any of those other states matter? It's going to delete the node, right?

Reply:
It doesn't, mostly for consistency sake

Reply:
OK. By the time execution gets to this point, there's really no turning back. The host object in k8s is going to be removed as soon as we release it. So I want to ensure that Ironic's database is cleaned up without a lot of fuss. There's a separate path for real deprovisioning still, so in the vast majority of cases we're going to follow the more complete and expected process.